[CK_TILE] Add indexing to pooling operator (Lwpck 3892)

amd-yashagar opened the pull request:
This PR extends the pooling kernel to optionally output the indices of selected elements during max and absolute max pooling operations. 

- Add IndexDataType template parameter to pooling problem and kernel definitions

- Enable pooling kernel to output indices of selected elements during max/absmax pooling

- Add overloaded operators for Max and AbsMax that track when values change using bool changed parameter

-  Support optional index buffer allocation and management in device memory

- Modify BlockReduce2d classes to handle index tensors alongside value tensors

-  Add separate shared memory allocation for index data in cross-warp reductions

- Create validate_pool_indices function to verify index correctness

- Modify pool3d.cpp example to demonstrate index output functionality

- Add tests for index output and 2D pooling


## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

amd-yashagar committed:
Add indexing support to pooling operator

- Add IndexDataType template parameter to pooling problem and kernel
definitions

- Enable pooling kernel to output indices of selected elements during
max/absmax pooling

- Add overloaded operators for Max and AbsMax that track when values
change using bool changed parameter

-  Support optional index buffer allocation and management in device
memory

- Modify BlockReduce2d classes to handle index tensors alongside value
tensors

-  Add separate shared memory allocation for index data in cross-warp
reductions

- Create validate_pool_indices function to verify index correctness

- Modify pool3d.cpp example to demonstrate index output functionality

- Add tests for index output

amd-yashagar committed:
fixes

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3892

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3892

aosewski commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -171,8 +181,15 @@ bool run(const ck_tile::ArgParser& arg_parser)
             in, out_ref, kernel_args, ReduceOp{});
         out_buf.FromDevice(out.mData.data());
         pass = ck_tile::check_err(out, out_ref);
+        std::cout << "value_valid:" << (pass ? "y" : "n") << std::flush << std::endl;
```
Flush is unnecessary. `std::endl` is already doing that. 
```suggestion
        std::cout << "value_valid:" << (pass ? "y" : "n") << std::endl;
```

aosewski commented on example/ck_tile/36_pooling/pool3d.cpp:
```diff
@@ -171,8 +181,15 @@ bool run(const ck_tile::ArgParser& arg_parser)
+            pass = ck_tile::validate_pool_indices(in, out, out_index);
 
-        std::cout << "valid:" << (pass ? "y" : "n") << std::flush << std::endl;
+            std::cout << "index_valid:" << (pass ? "y" : "n") << std::flush << std::endl;
```
```suggestion
            std::cout << "index_valid:" << (pass ? "y" : "n")  << std::endl;
```

aosewski commented on include/ck_tile/core/utility/reduce_operator.hpp:
```diff
@@ -85,6 +85,22 @@ struct Max
+              typename = std::enable_if_t<std::is_same_v<T, float> || std::is_same_v<T, double> ||
+                                          std::is_same_v<T, int32_t> || std::is_same_v<T, int8_t> ||
+                                          std::is_same_v<T, half_t> || std::is_same_v<T, bf16_t> ||
+                                          std::is_same_v<T, fp8_t> || std::is_same_v<T, bf8_t>>>
```
We have a utility function like `is_any_of<....>` which would make this code much more concise.

aosewski commented on include/ck_tile/core/utility/reduce_operator.hpp:
```diff
@@ -85,6 +85,22 @@ struct Max
+    CK_TILE_HOST_DEVICE constexpr T operator()(T& y, const T x, bool& changed) const
+    {
+        T new_max = max(y, x);
+        if(new_max > y)
```
Good catch.

aosewski commented on include/ck_tile/core/utility/reduce_operator.hpp:
```diff
@@ -108,6 +124,22 @@ struct AbsMax
+              typename = std::enable_if_t<std::is_same_v<T, float> || std::is_same_v<T, double> ||
+                                          std::is_same_v<T, int32_t> || std::is_same_v<T, int8_t> ||
+                                          std::is_same_v<T, half_t> || std::is_same_v<T, bf16_t> ||
+                                          std::is_same_v<T, fp8_t> || std::is_same_v<T, bf8_t>>>
```
Pls use `is_any_of`

aosewski commented on include/ck_tile/host/reference/reference_pool.hpp:
```diff
@@ -145,4 +146,46 @@ CK_TILE_HOST void reference_pool3d(const HostTensor<InDataType>& input,
+    // Validate each output position's index
+    for(int out_pos = 0; out_pos < static_cast<int>(output_size); out_pos++)
+    {
+        auto kernel_idx = output_index.mData[out_pos];
```
I think that HostTensor should have overloaded array access operator. 

```suggestion
        auto kernel_idx = output_index[out_pos];
```

aosewski commented on include/ck_tile/host/reference/reference_pool.hpp:
```diff
@@ -145,4 +146,46 @@ CK_TILE_HOST void reference_pool3d(const HostTensor<InDataType>& input,
+/// @param output_index Output indices tensor (to validate)
+/// @return true if all indices are valid, false otherwise
+template <typename InDataType, typename OutDataType, typename IndexDataType>
+CK_TILE_HOST bool validate_pool_indices(const HostTensor<InDataType>& input,
```
This isn't a good function. Just validating whether the index is somewhere between begin and end of tensor is not enough. 
Please add index calculation to above reference pooling. Then in example you may just call `check_err` as usual on both indexes and output values.

aosewski reviewed (changes requested):
Will finish tomorrow.

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3892

aosewski commented on include/ck_tile/ops/reduce/pipeline/reduce2d_default_policy.hpp:
```diff
@@ -76,5 +80,31 @@ struct Reduce2dDefaultPolicy
     }
+
+    template <typename Problem>
+    CK_TILE_HOST_DEVICE static constexpr index_t GetSmemSizeForIndices()
```
```suggestion
    CK_TILE_HOST_DEVICE static constexpr index_t GetIndicesSmemSize()
```

aosewski commented on include/ck_tile/ops/reduce/pipeline/reduce2d_default_policy.hpp:
```diff
@@ -76,5 +80,31 @@ struct Reduce2dDefaultPolicy
+        }
+        else
+        {
+            return 1; // zero size arrays are an extension
```
If you would have a simple function like `constexpr NeedIndicesSmem() -> bool` where you'd just put the condition `return Problem::kNeedCrossWarpSync && Problem::kOutputIndex` then you could guard smem allocation with it. 
[Update] Actually you already check those conditions in kernel logic, so I'd just guard smem allocation and use it if needed.

aosewski commented on include/ck_tile/ops/pooling/kernel/pool_kernel.hpp:
```diff
@@ -387,16 +424,54 @@ struct PoolKernel
+            }
+
+            block_reduce2d_sync(y_tile, y_index_tile, reduce_op);
+            block_reduce2d_cross_warp(y_tile, y_index_tile, smem, smem_indices, reduce_op);
```
Please guard this with `Problem::kNeedCrossWarpSync` and put here smem allocation, then you won't have a problem with empty smem arrays.

aosewski commented on include/ck_tile/ops/pooling/pipeline/pool_default_policy.hpp:
```diff
@@ -76,5 +80,32 @@ struct PoolDefaultPolicy
     }
+
+    template <typename Problem>
+    CK_TILE_HOST_DEVICE static constexpr index_t GetSmemSizeForIndices()
```
I think it might be better that instead of guarding here, you should guard the smem allocation and here always return the required amount of mem.

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -61,39 +121,59 @@ struct BlockReduce2d
+                    ReducePacksPerXDim{});
+    }
 
-#if 0
-        constexpr auto I0 = number<0>{};
-        constexpr auto I1 = number<1>{};
-        constexpr auto spans = XDistributedTensor_::get_distributed_spans();
```
I'd leave this commented out code - but maybe change it to comment. It is really useful to understand what's going on.

amd-yashagar replied:
re-added the comment

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -61,39 +121,59 @@ struct BlockReduce2d
-            },
-            ReducePacksPerXDim{});
+        EmptyIndexTensor empty_index_tensor{};
+        EmptyIndexCalculator empty_index_calc{};
```
What about having those as default values in `reduce_impl` ?

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -50,6 +50,66 @@ struct BlockReduce2d
+
+    template <typename XDistributedTensor_,
+              typename YDistributedTensor_,
+              typename YIndexDistributedTensor_,
```
```suggestion
              typename YIndexDistributedTensor_, = EmptyIndexTensor
```

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -50,6 +50,66 @@ struct BlockReduce2d
+              typename YDistributedTensor_,
+              typename YIndexDistributedTensor_,
+              typename ReduceFunc,
+              typename IndexCalculatorFunc,
```
```suggestion
              typename IndexCalculatorFunc = EmptyIndexCalculator,
```

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -50,6 +50,66 @@ struct BlockReduce2d
+              typename ReducePacksPerXDim>
+    CK_TILE_DEVICE void reduce_impl(const XDistributedTensor_& x_tensor,
+                                    YDistributedTensor_& y_tensor,
+                                    YIndexDistributedTensor_& y_index_tensor,
```
```suggestion
                                    YIndexDistributedTensor_& y_index_tensor = YIndexDistributedTensor_{},
```

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -50,6 +50,66 @@ struct BlockReduce2d
+                                    YDistributedTensor_& y_tensor,
+                                    YIndexDistributedTensor_& y_index_tensor,
+                                    const ReduceFunc& reduce_func,
+                                    const IndexCalculatorFunc& index_calculator,
```
```suggestion
                                    const IndexCalculatorFunc& index_calculator = IndexCalculatorFunc{},
```

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -252,15 +418,71 @@ struct BlockReduce2dCrossWarpSync
+
+    private:
+    // Empty type for when index tracking is not needed
+    struct EmptyIndexTensor
```
Let's put this code to the anonymous namespace outside the reduction structures in this file, so it could be reused.

aosewski commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -276,35 +498,111 @@ struct BlockReduce2dCrossWarpSync
+                    }
+                });
+            });
+            block_sync_lds(); // TODO: we don't need sync here
```
Is this comment still valid?

aosewski reviewed (changes requested):
Good work, but few things need attention.

aosewski committed:
Refactor BlockReduce2D functions to get rid auxiliary private types.

aosewski committed:
Merge remote-tracking branch 'origin/develop' into LWPCK-3892

amd-yashagar committed:
comment resolutions and some changes to block_reduce2d

- index reference implementation improved
- reduce_operator.hpp cleanedup
- updated the block_reduce2d.hpp to have index calculation for
BlockReduce2dLinearCrossWarpSync as well

amd-yashagar committed:
conditionally used variable declaration improvement

- the conditionally used vairbales are used only when indexing is
enabled. To inform the compiler that they may be unused and declare them
with least size possible. This may allow it to be optimized compared to
the previous declarations

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3892

amd-yashagar committed:
comment resolutions

amd-yashagar committed:
lexical ordering of the indicies

- introduced accumulate methods that handle the intermediate steps if
needed to order the indexes

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3892

amd-yashagar committed:
add reduce_operator_accumulate.hpp to core.hpp
